source/server/listener_manager_impl.cc

```diff
@@ -289,6 +304,16 @@ bool ListenerManagerImpl::addOrUpdateListener(const Json::Object& json) {
   return true;
 }
 
+bool ListenerManagerImpl::hasListenerWithAddress(const ListenerList& list,
+                                                 const Network::Address::Instance& address) {
+  for (const auto& listener : list) {
```

Comment:
Kind of unfortunate that this requires walking the list and the non-error case requires going all the way to the end with a bunch of cache unfriendly pointer chasing. Since this is only run on listener update maybe it's not really a big performance issue. Would changing ListenerList into a vector fix some of these issues?